File: Assets/IFramework/UI/MVVM/UIView.cs
/*********************************************************************************
 *Author:         OnClick
 *Version:        0.0.1.440
 *UnityVersion:   2018.4.17f1
 *Date:           2020-02-28
 *Description:    IFramework
 *History:        2018.11--
*********************************************************************************/
using IFramework.Modules.Message;
using IFramework.Modules.MVVM;
using System;
using UnityEngine.Events;
using UnityEngine.UI;
using static IFramework.UI.MvvmGroups;

namespace IFramework.UI
{
    /// <summary>
    /// UIView 基类
    /// </summary>
    public abstract class UIView : View, IViewStateEventHandler
    {

        public UIPanel panel;
        private ViewEventType _lastState = ViewEventType.None;

        public ViewEventType lastState { get { return _lastState; } }
        protected void Publish(IEventArgs args)
        {
            this.message.Publish(this.GetType(), args);
        }
        protected virtual void Show()
        {
            panel.Show();
        }
        protected virtual void Hide()
        {
            panel.Hide();
        }
        public virtual void Pause()
        {
            panel.Pause();
        }
        public virtual void UnPause()
        {
            panel.UnPause();
        }

        protected UIView Bind(Text text, Func<string> getter)
        {
            handler.BindProperty(() => {
                string tmp = getter();
                if (tmp != text.text)
                {
                    text.text = tmp;
                }
            });
            return this;
        }
        protected UIView Bind(InputField input, Func<string> getter)
        {
            handler.BindProperty(() => {
                string tmp = getter();
                if (tmp != input.text)
                {
                    input.text = tmp;
                }
            });
            return this;
        }
        protected UIView Bind(Slider slider, Func<float> getter)
        {
            handler.BindProperty(() => {
                float tmp = getter();
                if (slider.value != tmp)
                {
                    slider.value = tmp;
                }
            });
            return this;
        }
        protected UIView Bind(Toggle toggle, Func<bool> getter)
        {
            handler.BindProperty(() => {
                bool tmp = getter();
                if (tmp != toggle.isOn)
                {
                    toggle.isOn = tmp;
                }
            });
            return this;
        }

        protected UIView BindInputField(InputField input, UnityAction<string> callback)
        {
            input.onValueChanged.AddListener(callback);
            return this;
        }
        protected UIView BindToggle(Toggle toggle, UnityAction<bool> callback)
        {
            toggle.onValueChanged.AddListener(callback);
            return this;
        }
        protected UIView BindSlider(Slider slider, UnityAction<float> callback)
        {
            slider.onValueChanged.AddListener(callback);
            return this;
        }
        protected UIView BindOnEndEdit(InputField input, UnityAction<string> callback)
        {
            input.onEndEdit.AddListener(callback);
            return this;
        }
        protected UIView BindOnValidateInput(InputField input, InputField.OnValidateInput callback)
        {
            input.onValidateInput = callback;
            return this;
        }
        protected UIView BindButton(Button button, UnityAction callback)
        {
            button.onClick.AddListener(callback);
            return this;
        }


        protected abstract void OnLoad();
        protected abstract void OnTop(UIEventArgs arg);
        protected abstract void OnPress(UIEventArgs arg);
        protected abstract void OnPop(UIEventArgs arg);
        protected abstract void OnClear();

        protected abstract void OnShow();
        protected abstract void OnHide();
        protected abstract void OnPause();
        protected abstract void OnUnPause();
        protected abstract void OnClose();

        void IViewStateEventHandler.OnLoad()
        {
            _lastState = ViewEventType.OnLoad;
            OnLoad();
        }
        void IViewStateEventHandler.OnTop(UIEventArgs arg)
        {
            _lastState = ViewEventType.OnTop;
            OnTop(arg);
        }
        void IViewStateEventHandler.OnPress(UIEventArgs arg)
        {
            _lastState = ViewEventType.OnPress;
            OnPress(arg);
        }
        void IViewStateEventHandler.OnPop(UIEventArgs arg)
        {
            _lastState = ViewEventType.OnPop;
            OnPop(arg);
        }
        void IViewStateEventHandler.OnClear()
        {
            _lastState = ViewEventType.OnClear;
            OnClear();
        }

        void IViewStateEventHandler.OnShow()
        {
            OnShow();
        }

        void IViewStateEventHandler.OnHide()
        {
            OnHide();
        }

        void IViewStateEventHandler.OnPause()
        {
            OnPause();
        }

        void IViewStateEventHandler.OnUnPause()
        {
            OnUnPause();
        }

        void IViewStateEventHandler.OnClose()
        {
            OnClose();
        }
    }
    public abstract class UIView<VM, Panel> : UIView where VM : UIViewModel where Panel : UIPanel
    {
        public VM Tcontext { get { return this.context as VM; } }
        public Panel Tpanel { get { return this.panel as Panel; } }
    }
}

File: Assets/IFramework/HotFix/Lua/Resources/UI/UIView.lua
---@class UIView
local UIView = class("UIView")


---@param message Delegate 委托
---@param context ViewModel ViewModel 
---@param panel CS.IFramework.UI.UIPanel UI
function UIView:ctor(message, context, panel)
    self.message = message
    self.context = context
    self.panel = panel
    self:OnLoad()
    self:BindProperty()
end

--- @type 绑定对应ViewModel的字段
---@param field string 字段名字
---@param func function 回调
function UIView:BindContextField(field, func)
    local handle = Handler(self, func)
    self.context:Subscribe(field, handle)
end

--- @type 发布一个事件给对应的ViewModel
---@param code any 用于区分的事件编号
---@param ... any 其他参数
function UIView:PublishViewEvent(code, ...)
    self.message:Invoke(code, ...)
end
--- 以下是查找物体

--- @type 根据指定路径查找组件
---@param path string 相对路径
---@param type Component 组件类型
function UIView:GetComponent(path, type)
    local transform = self.panel.transform:Find(path)
    if not transform then
        error("can't find transfom with " .. path .. " in panel " .. self.panel.name)
    end
    return transform:GetComponent(type) or transform:GetComponentInChildren(type, true)
end

--- 以下是绑定ui事件

--- @type 绑定对应Slider的onValueChanged方法
---@param Slider Slider 对应的Slider
---@param func function 回调
function UIView:BindSlider(Slider, func)
    local handle = Handler(self, func)
    Slider.onValueChanged:AddListener(handle)
end

--- @type 取消绑定对应Slider的onValueChanged方法
---@param Slider Slider 对应的Slider
function UIView:UnBindToggle(Slider)
    Slider.onValueChanged:RemoveAllListeners()
    Slider.onValueChanged:Invoke()
end

--- @type 绑定对应Toggle的onValueChanged方法
---@param Toggle Toggle 对应的Toggle
---@param func function 回调
function UIView:BindToggle(Toggle, func)
    local handle = Handler(self, func)
    Toggle.onValueChanged:AddListener(handle)
end

--- @type 取消绑定对应Toggle的onValueChanged方法
---@param Toggle Toggle 对应的Toggle
function UIView:UnBindToggle(Toggle)
    Toggle.onValueChanged:RemoveAllListeners()
    Toggle.onValueChanged:Invoke()
end

--- @type 绑定对应InputField的onValueChanged方法
---@param InputField InputField 对应的InputField
---@param func function 回调
function UIView:BindInputField(InputField, func)
    local handle = Handler(self, func)
    InputField.onValueChanged:AddListener(handle)
end

--- @type 取消绑定对应InputField的onValueChanged方法
---@param InputField InputField 对应的InputField
function UIView:UnBindInputField(InputField)
    InputField.onValueChanged:RemoveAllListeners()
    InputField.onValueChanged:Invoke()
end

--- @type 绑定对应InputField的onEndEdit方法
---@param InputField InputField 对应的InputField
---@param func function 回调
function UIView:BindOnEndEdit(InputField, func)
    local handle = Handler(self, func)
    InputField.onEndEdit:AddListener(handle)
end

--- @type 取消绑定对应InputField的onEndEdit方法
---@param InputField InputField 对应的InputField
function UIView:UnBindOnEndEdit(InputField)
    InputField.onEndEdit:RemoveAllListeners()
    InputField.onEndEdit:Invoke()
end

--- @type 绑定对应InputField的OnValidateInput方法
---@param InputField InputField 对应的InputField
---@param func function 回调
function UIView:BindOnValidateInput(InputField, func)
    local handle = Handler(self, func)
    InputField.onValidateInput = handle
end

--- @type 取消绑定对应InputField的OnValidateInput方法
---@param InputField InputField 对应的InputField
function UIView:UnBindOnValidateInput(InputField)
    InputField.onValidateInput = nil
end

--- @type 绑定对应Button的onClick方法
---@param btn Button 对应的Button
---@param func function 回调
function UIView:BindButton(btn, func)
    local handle = Handler(self, func)
    btn.onClick:AddListener(handle)
end

--- 以下是绑定ui事件
--- @type 取消绑定对应Button的onClick方法
---@param btn Button 对应的Button
function UIView:UnBindButton(btn)
    btn.onClick:RemoveAllListeners()
    btn.onClick:Invoke()
end

--- 以下方法来自UIPanel

function UIView:Show()
    self.panel:Show()
end
function UIView:Hide()
    self.panel:Hide()
end
function UIView:Pause()
    self.panel:Pause()
end
function UIView:UnPause()
    self.panel:UnPause()
end
function UIView:Close()
    self.panel:Close()
end

--- 以下是回调

function UIView:BindProperty()
end
function UIView:Dispose()
end
function UIView:OnLoad()
end
function UIView:OnTop(arg)
end
function UIView:OnPress(arg)
end
function UIView:OnPop(arg)
end
function UIView:OnClear()
end

function UIView:OnShow()
end

function UIView:OnHide()
end

function UIView:OnPause()
end

function UIView:OnUnPause()
end

function UIView:OnClose()
end

return UIView
